README.md

# rate_in_store

A Flutter rate plugin.

# Install

```
dependencies:
  rate_in_store: ^0.0.1
```

# How to use

```dart
RateInStore.rate(iOSAppID: 'id2343214123', androidAppID: 'com.xxx.xxx');
```

# On iOS

```swift
if #available(iOS 10.3, *) {
    SKStoreReviewController.requestReview()
} else {
    guard let appID = call.arguments as? String else {
        print("appid cannot be null")
        return
    }
    
    var components = URLComponents(url: URL(string: "https://itunes.apple.com/app/\(appID)")!, resolvingAgainstBaseURL: false)
    components?.queryItems = [
        URLQueryItem(name: "action", value: "write-review")
    ]
    guard let writeReviewURL = components?.url else {
        return
    }
    
    if #available(iOS 10.0, *) {
        UIApplication.shared.open(writeReviewURL)
    } else {
        UIApplication.shared.openURL(writeReviewURL)
    }
}
```

# On Android

Use [url_launch](https://pub.dev/packages/url_launcher)

```dart
launch('market://details?id=$appID');
```

CHANGELOG.md
## 0.0.1 [01.08.2020]
* add rate method


test/rate_in_store_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:rate_in_store/rate_in_store.dart';

void main() {
  const MethodChannel channel = MethodChannel('rate_in_store');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await RateInStore.platformVersion, '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:rate_in_store_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:rate_in_store/rate_in_store.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      platformVersion = await RateInStore.platformVersion;
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on: $_platformVersion\n'),
        ),
      ),
    );
  }
}


android/src/main/kotlin/com/hellotalk/rate_in_store/RateInStorePlugin.kt
package com.hellotalk.rate_in_store

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

/** RateInStorePlugin */
public class RateInStorePlugin: FlutterPlugin, MethodCallHandler {
  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "rate_in_store")
    channel.setMethodCallHandler(RateInStorePlugin());
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar) {
      val channel = MethodChannel(registrar.messenger(), "rate_in_store")
      channel.setMethodCallHandler(RateInStorePlugin())
    }
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
  }
}


ios/Classes/SwiftRateInStorePlugin.swift
import Flutter
import UIKit
import StoreKit

public class SwiftRateInStorePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "rate_in_store", binaryMessenger: registrar.messenger())
    let instance = SwiftRateInStorePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "rate" {
        if #available(iOS 10.3, *) {
            SKStoreReviewController.requestReview()
        } else {
            // id557130558
            guard let appID = call.arguments as? String else {
                print("appid cannot be null")
                return
            }
            
            var components = URLComponents(url: URL(string: "https://itunes.apple.com/app/\(appID)")!, resolvingAgainstBaseURL: false)
            components?.queryItems = [
                URLQueryItem(name: "action", value: "write-review")
            ]
            guard let writeReviewURL = components?.url else {
                return
            }
            
            if #available(iOS 10.0, *) {
                UIApplication.shared.open(writeReviewURL)
            } else {
                UIApplication.shared.openURL(writeReviewURL)
            }
        }
    }
  }
}
